File: Array/3036. Number of Subarrays That Match a Pattern II.cpp
/*
3036. Number of Subarrays That Match a Pattern II
*/

// TC:O(m+n) SC:O(m+n)
class Solution {
public:
    int countMatchingSubarrays(vector<int>& nums, vector<int>& pattern) {
        string A;
        for(int i = 1; i < nums.size(); i++){
            if(nums[i] > nums[i-1]){
                A.push_back('A');
            }
            else if(nums[i] == nums[i-1]){
                A.push_back('B');
            }
            else A.push_back('C');
        }
        
        string B;
        for(auto x : pattern){
            if(x == 1) B.push_back('A');
            else if(x == 0) B.push_back('B');
            else B.push_back('C');
        }
        // 看pattern跟arr有多少個match
        vector<int> suf = preprocessor(B);
        
        int n = A.size();
        int m = B.size();
        if(n < m) return 0;
        
        int res = 0;
        vector<int> dp(n);
        dp[0] = (B[0] == A[0]);
        
        if(m == 1 && dp[0] == 1) res += 1;
        
        for(int i = 1; i < n; i++){
            int j = dp[i-1];
            while(j > 0 && A[i] != B[j]){
                j = suf[j-1];
            }
            
            dp[i] = j + (A[i] == B[j]);
            if(dp[i] == m){
                res += 1;
            }
        }
        
        return res;
    }
    vector<int> preprocessor(string& s){
        int n = s.size();
        vector<int> suffix(n,0);
        suffix[0] = 0;
        for(int i = 1; i<n; i++){
            int j = suffix[i-1];
            while(j > 0 && s[i] != s[j]){
                j = suffix[j-1];
            }
            suffix[i] = j + (s[i] == s[j]);
        }

        return suffix;
    }
};

/*
此題再說 pattern 如果某index 數值為 1 nums的此數值+1的所有倍數
m = 3 pattern = [1,1]
ex : k = 1 都必須後面大於前面
k= 0 都必須兩兩相等
k = -1 都必須後面小於前面
藉由上述三個規則和出的subarray有幾個
[1 0 -1 1 1 0 -1]
*/

File: Array/3036. Number of Subarrays That Match a Pattern II.py
#3036. Number of Subarrays That Match a Pattern II
# TC:O(m+n) SC:O(m+n)

class Solution:
    def countMatchingSubarrays(self, nums: List[int], s: List[int]) -> int:
       
        m = len(s)
        # 可以讓原本的list 直接多append一段list, 跟 append差別就在append只能新增一個
        # pairwise 就是直接從可迭代的資料結構取相鄰的元素
        s.extend((y>x) - (y<x) for x, y in pairwise(nums))

        n = len(s)
        z = [0]*n
        l = r = 0
        for i in range(1,n):
            if i <= r:
                z[i] = min(z[i-l], r-i+1) # 看圖取往回推的比較點z-value or 看z-box到當下pos i的當前長度是多少 取最短
            while i + z[i] < n and s[z[i]] == s[z[i]+i]:
                l,r = i, i+z[i]
                z[i] += 1
        return sum(lcp >= m for lcp in z[m:])
